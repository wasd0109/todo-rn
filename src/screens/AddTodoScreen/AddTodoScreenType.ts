import { ParamListBase, Route } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";

export type AddTodoScreenProps = {
  navigation: StackNavigationProp<ParamListBase, string>;
};
