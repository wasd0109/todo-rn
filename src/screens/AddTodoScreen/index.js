import AddTodoScreen from "./AddTodoScreen";
export default AddTodoScreen;
