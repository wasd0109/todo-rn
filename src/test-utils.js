import React from "react";
import "react-native-paper";
import { render } from "@testing-library/react-native";
import { Provider as PaperProvider } from "react-native-paper";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";

const AllTheProviders = ({ children }) => {
  return <PaperProvider>{children}</PaperProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react-native";

// override render method
export { customRender as render };

const Stack = createStackNavigator();
const MockedNavigator = ({ screenName, component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name={screenName}
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockedNavigator;
