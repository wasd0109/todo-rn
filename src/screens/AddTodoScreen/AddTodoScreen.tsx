import React, { useState } from "react";
import { StyleSheet, Text, View } from "react-native";
import { Button, TextInput } from "react-native-paper";
import Spacer from "../../components/Spacer/Spacer";
import { useDispatch } from "react-redux";
import { addTodo } from "../../slices/todoSlices";
import { AddTodoScreenProps } from "./AddTodoScreenType";
import generateTodo from "../../utils/generateTodo";

const TODO_TITLE_PLACEHOLDER = "Buy groceries";

const AddTodoScreen = ({ navigation }: AddTodoScreenProps) => {
  const [title, setTitle] = useState("");
  const [error, setError] = useState("");
  const dispatch = useDispatch();
  return (
    <View style={styles.containerStyle}>
      <Spacer>
        <TextInput
          label="Title"
          mode="outlined"
          placeholder={TODO_TITLE_PLACEHOLDER}
          testID="title-input"
          value={title}
          onChangeText={setTitle}
        />
      </Spacer>
      {error ? (
        <Spacer>
          <Text style={styles.errorStyle}>{error}</Text>
        </Spacer>
      ) : null}
      <Spacer>
        <Button
          mode="contained"
          onPress={() => {
            if (!title) return setError("Please enter title");
            setError("");
            dispatch(addTodo(generateTodo(title)));
            navigation.navigate("Home");
          }}
        >
          Add Todo
        </Button>
      </Spacer>
    </View>
  );
};

export default AddTodoScreen;

const styles = StyleSheet.create({
  containerStyle: {
    marginHorizontal: 15,
    justifyContent: "center",
    alignContent: "center",
    flex: 1,
  },
  errorStyle: {
    color: "red",
  },
});
