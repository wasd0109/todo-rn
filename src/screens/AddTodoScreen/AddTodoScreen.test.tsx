import React from "react";
import { render, fireEvent } from "../../test-utils";
import AddTodoScreen from "./AddTodoScreen";
import { RenderAPI } from "@testing-library/react-native";

import { store } from "../../store";
import { AddTodoScreenProps } from "./AddTodoScreenType";
import { ParamListBase, Route } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";

const TODO_TITLE_PLACEHOLDER = "Buy groceries";
const TEST_TODO_TITLE = "Buy things";

let component: RenderAPI;

let props = {
  navigation: {
    navigate: jest.fn(),
  },
} as any;

beforeEach(() => {
  component = render(<AddTodoScreen {...props} />);
});

describe("AddTodoScreen render correctly", () => {
  test("AddTodoScreen render", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Todo title input render correctly", () => {
    const { getAllByTestId } = component;
    expect(getAllByTestId("title-input").length).toEqual(1);
  });

  test("Add Todo button render correctly", () => {
    const { getAllByText } = component;
    expect(getAllByText("Add Todo").length).toEqual(1);
  });
});

describe("AddTodoScreen function properly", () => {
  test("Todo title input change value on input", () => {
    const { getByTestId, getAllByDisplayValue } = component;
    const titleInputRef = getByTestId("title-input");
    fireEvent.changeText(titleInputRef, TEST_TODO_TITLE);
    expect(getAllByDisplayValue(TEST_TODO_TITLE).length).toEqual(1);
  });

  test("Add Todo button return to home on press when title was entered", () => {
    let props = {
      navigation: {
        navigate: jest.fn(),
      },
    } as any;
    const { getByText, getByTestId } = render(<AddTodoScreen {...props} />);
    const buttonRef = getByText("Add Todo");
    const titleInputRef = getByTestId("title-input");
    fireEvent.changeText(titleInputRef, TEST_TODO_TITLE);
    fireEvent.press(buttonRef);
    expect(props.navigation.navigate).toBeCalledWith("Home");
  });

  test("Add Todo button does not return to home when title was empty", () => {
    let props = {
      navigation: {
        navigate: jest.fn(),
      },
    } as any;
    const { getByText } = render(<AddTodoScreen {...props} />);
    const buttonRef = getByText("Add Todo");
    fireEvent.press(buttonRef);
    expect(props.navigation.navigate).toBeCalledTimes(0);
  });

  test("Add Todo button add Todo to state", () => {
    const { getByText, getByTestId } = render(<AddTodoScreen {...props} />);
    const buttonRef = getByText("Add Todo");
    const titleInputRef = getByTestId("title-input");
    fireEvent.changeText(titleInputRef, TEST_TODO_TITLE);
    fireEvent.press(buttonRef);
    expect(store.getState().todo.todoList[0].title).toEqual(TEST_TODO_TITLE);
  });
});
