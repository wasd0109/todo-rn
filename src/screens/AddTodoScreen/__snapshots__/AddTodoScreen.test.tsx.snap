// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoScreen render correctly AddTodoScreen render 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 15,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={Object {}}
      >
        <View>
          <View
            pointerEvents="none"
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                Object {
                  "backgroundColor": "#f6f6f6",
                  "borderColor": "rgba(0, 0, 0, 0.54)",
                  "borderRadius": 4,
                  "borderWidth": 1,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 0,
                },
                Object {
                  "minHeight": 64,
                  "paddingTop": 8,
                },
              ]
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "#f6f6f6",
                    "bottom": 4,
                    "left": 10,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                    ],
                    "width": 8,
                  }
                }
              />
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "#f6f6f6",
                    "color": "transparent",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 18,
                    "opacity": 0,
                    "paddingHorizontal": 0,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 37,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                      Object {
                        "scaleY": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                Title
              </Text>
              <Text
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "#6200ee",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingHorizontal": 14,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 36,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                Title
              </Text>
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingHorizontal": 14,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 36,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                Title
              </Text>
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder=""
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              rejectResponderTermination={true}
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "paddingHorizontal": 14,
                    "zIndex": 1,
                  },
                  Object {
                    "height": 56,
                  },
                  Object {
                    "paddingBottom": 0,
                    "paddingTop": 0,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  Array [
                    Object {},
                  ],
                ]
              }
              testID="title-input"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#6200ee",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Add Todo
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
